### PR TITLE
Gql optimize builder

### PR DESCRIPTION
**Pull request by asmab89**

### CODE REVIEW

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,1011 @@
...
+                    KindOfFieldName = element.fieldType,
+                    kindOfFieldType = element.kindOfFieldType,
+                    typeName = element.fieldName,
+                    isKindOfFieldTypeOptional = element.isKindOfFieldTypeOptional,
+                    isKindOfFieldNameOptional = element.isKindOfFieldNameOptional,
+                    enumValues = element.enumValues,
+                    unionTypes = element.unionTypes,
+                    interfaceTypes = element.interfaceTypes,
+                    isFieldNameWithArgs = element.isFieldNameWithArgs,
+                )
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:505]: aren't some of these assignments on same variable names redundant? eg, `interfaceTypes = element.interfaceTypes`

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,1011 @@
...
+                            KindOfFieldName = element.kindOfFieldType,
+                            kindOfFieldType = element.kindOfFieldType,
+                            typeName = element.fieldName,
+                            isKindOfFieldTypeOptional = element.isKindOfFieldTypeOptional,
+                            isKindOfFieldNameOptional = element.isKindOfFieldNameOptional,
+                            enumValues = element.enumValues,
+                            unionTypes = element.unionTypes,
+                            interfaceTypes = element.interfaceTypes,
+                            isFieldNameWithArgs = element.isFieldNameWithArgs
+                        )
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:534]: see previous comment

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,1011 @@
...
+            KindOfFieldName = element.KindOfFieldName,
+            kindOfFieldType = element.kindOfFieldType,
+            typeName = element.fieldType,
+            isKindOfFieldTypeOptional = secondElement.isKindOfFieldTypeOptional,
+            isKindOfFieldNameOptional = secondElement.isKindOfFieldNameOptional,
+            enumValues = element.enumValues,
+            unionTypes = element.unionTypes,
+            interfaceTypes = element.interfaceTypes,
+            isFieldNameWithArgs = element.isFieldNameWithArgs
+        )
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:558]: see previous comment. check all of theses cases in this file. ie, if copied value is not changed, no need to manually assign it, isn't it?

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt [hunk 4]
```diff
@@ -0,0 +1,1011 @@
...
+                KindOfFieldName = tableElement.KindOfFieldName,
+                kindOfFieldType = tableElement.kindOfFieldType,
+                typeName = tableElement.fieldType,
+                isKindOfFieldTypeOptional = isKindOfTableFieldTypeOptional,
+                isKindOfFieldNameOptional = isKindOfTableFieldOptional,
+                enumValues = tableElement.enumValues,
+                unionTypes = tableElement.unionTypes,
+                interfaceTypes = tableElement.interfaceTypes,
+                isFieldNameWithArgs = tableElement.isFieldNameWithArgs
+            )
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:825]: see previous comments

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt [hunk 5]
```diff
@@ -0,0 +1,1011 @@
...
+        for (elementInUnionTypes in element.unionTypes) {//Browse all objects defining the union
+            accum += 1
+            if (checkDepth(accum, maxNumberOfGenes)) {
+                history.addLast(elementInUnionTypes)
+                val copy = element.copy(
+                    typeName = elementInUnionTypes,
+                    kindOfFieldType = element.kindOfFieldType,
+                    isKindOfFieldTypeOptional = element.isKindOfFieldTypeOptional,
+                    isKindOfFieldNameOptional = element.isKindOfFieldNameOptional,
+                    fieldName = elementInUnionTypes
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:900]: see previous comments

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt [hunk 6]
```diff
@@ -0,0 +1,1011 @@
...
+        for (elementInInterfaceTypes in element.interfaceTypes) {//Browse all additional objects in the interface
+            accum += 1
+            if (checkDepth(accum, maxNumberOfGenes)) {
+                history.addLast(elementInInterfaceTypes)
+                val copy = element.copy(
+                    typeName = elementInInterfaceTypes,
+                    kindOfFieldType = element.kindOfFieldType,
+                    isKindOfFieldTypeOptional = element.isKindOfFieldTypeOptional,
+                    isKindOfFieldNameOptional = element.isKindOfFieldNameOptional,
+                    fieldName = elementInInterfaceTypes
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:941]: see previous comments

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt [hunk 7]
```diff
@@ -0,0 +1,1011 @@
...
+                )
+                val objGeneTemplate = createObjectGene(state, history, accum, maxNumberOfGenes, copy)
+
+
+                history.removeLast()
+
+                var myObjGne = objGeneTemplate as ObjectGene//To object
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:948]: hmmm... shouldn't then `createObjectGene` have a `ObjectGene` as return type?

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/StateBuilder.kt [hunk 8]
```diff
@@ -0,0 +1,1146 @@
...
+                                isKindOfFieldTypeOptional = true,
+
+                                kindOfFieldType = kind0.toString(),
+                                fieldType = elementInfields.type.name,
+                                typeName = elementInTypes.name,
+
+                                enumValues = listOf(),
+                                unionTypes = listOf(),
+                                interfaceTypes = listOf()
+                            )
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/StateBuilder.kt:415]: i guess these `listOf()` could be good option for default values in `Table`, isn't it?

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/Table.kt [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+        /**
+         * Describing if the field name has arguments
+         */
+        val isFieldNameWithArgs: Boolean = false,
+
+        /**
+         * Containing the enum values
+         */
+        val enumValues: List<String>,
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/Table.kt:59]: can put defaults values here, eg ` = listOf()`

**Review by arcuri82 — CHANGES_REQUESTED** [review]

**Comment by arcuri82** [comment]
@asmab89 Hi, before making new commits here, make sure to pull from `master` (i disabled C# on CI, which will make CI much faster now)

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt (hunk 1 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:498]: Done

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt (hunk 2 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:520]: Done

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt (hunk 3 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:537]: Done

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt (hunk 4 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:797]: Done

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt (hunk 5 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:869]: Done

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt (hunk 6 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt:907]: Done

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt (hunk 7 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/GraphQLActionBuilder.kt]: Done

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/StateBuilder.kt (hunk 8 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/StateBuilder.kt:385]: Done

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/Table.kt (hunk 9 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/problem/graphql/builder/Table.kt]: Done

**Pull request merged**